libkbfs/disk_block_cache_test.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"context"
+	"testing"
+
+	"github.com/keybase/kbfs/kbfsblock"
+	"github.com/keybase/kbfs/kbfscrypto"
+	"github.com/keybase/kbfs/tlf"
+	"github.com/stretchr/testify/require"
+	"github.com/syndtr/goleveldb/leveldb/errors"
+	"github.com/syndtr/goleveldb/leveldb/storage"
+)
+
+const (
+	testDiskBlockCacheMaxBytes uint64 = 1 << 20
+)
+
+type testDiskBlockCacheConfig struct {
+	codecGetter
+	logMaker
+}
+
+func newTestDiskBlockCacheConfig(t *testing.T) testDiskBlockCacheConfig {
+	return testDiskBlockCacheConfig{
+		newTestCodecGetter(),
+		newTestLogMaker(t),
+	}
+}
+
+func newDiskBlockCacheStandardForTest(config diskBlockCacheConfig,
+	maxBytes uint64) (*DiskBlockCacheStandard, error) {
+	blockStorage := storage.NewMemStorage()
+	lruStorage := storage.NewMemStorage()
+	return newDiskBlockCacheStandardFromStorage(config, blockStorage,
+		lruStorage, maxBytes)
+}
+
+func initDiskBlockCacheTest(t *testing.T) (*DiskBlockCacheStandard,
+	testDiskBlockCacheConfig) {
+	config := newTestDiskBlockCacheConfig(t)
+	cache, err := newDiskBlockCacheStandardForTest(config,
+		testDiskBlockCacheMaxBytes)
+	require.NoError(t, err)
+	return cache, config
+}
+
+func shutdownDiskBlockCacheTest(cache DiskBlockCache) {
+	cache.Shutdown()
+}
+
+func setupBlockForDiskCache(t *testing.T, config diskBlockCacheConfig) (
+	kbfsblock.ID, []byte, kbfscrypto.BlockCryptKeyServerHalf) {
+	ptr := makeRandomBlockPointer(t)
+	block := makeFakeFileBlock(t, false)
+	blockEncoded, err := config.Codec().Encode(block)
+	require.NoError(t, err)
+	serverHalf, err := kbfscrypto.MakeRandomBlockCryptKeyServerHalf()
+	require.NoError(t, err)
+	return ptr.ID, blockEncoded, serverHalf
+}
+
+func TestDiskBlockCachePutAndGet(t *testing.T) {
+	t.Parallel()
+	t.Log("Test that basic disk cache Put and Get operations work.")
+	cache, config := initDiskBlockCacheTest(t)
+	defer shutdownDiskBlockCacheTest(cache)
+
+	tlf1 := tlf.FakeID(0, false)
+	block1Id, block1Encoded, block1ServerHalf := setupBlockForDiskCache(t, config)
+
+	ctx := context.Background()
+
+	t.Log("Put a block into the cache.")
+	err := cache.Put(ctx, tlf1, block1Id, block1Encoded, block1ServerHalf)
+	require.NoError(t, err)
+	putTime, err := cache.getLru(tlf1, block1Id)
+	require.NoError(t, err)
+
+	t.Log("Get that block from the cache. Verify that it's the same.")
+	buf, serverHalf, err := cache.Get(ctx, tlf1, block1Id)
+	require.NoError(t, err)
+	require.Equal(t, block1ServerHalf, serverHalf)
+	require.Equal(t, block1Encoded, buf)
+
+	t.Log("Verify that the Get updated the LRU time for the block.")
+	getTime, err := cache.getLru(tlf1, block1Id)
+	require.NoError(t, err)
+	require.True(t, getTime.After(putTime))
```

Comment:
Maybe not on Windows XP: https://github.com/golang/go/issues/11313

Haha.  Ok.  I'd still prefer using a fake clock just in case, but I won't fight you on it if you don't like the idea.